### DISCUSSION TITLE
How to change the color of sign of breakpoint

### DISCUSSION THREAD

**Discussion by h4kuy4** [mfussenegger/nvim-dap#355] [Q&A]
I want to change the sign of breakpoint, and I have changed the sign, but don't how to change it color.
![dap1](https://user-images.githubusercontent.com/53516842/142823790-e33dd6dd-fb55-420f-bf63-862e7bb4973c.png)
I want to make the sign of breakpoint in the picture into red.

**Comment by mfussenegger** [comment]
Something like this should do the trick, depends on your highlight groups.

```
:lua vim.fn.sign_define('DapBreakpoint', {text='•', texthl='red', linehl='', numhl=''})
```

See `:help sign_define` the `SIGNS CONFIGURATION` section in `:h dap.txt`

**Reply by josefvn** [reply]
I would like to do the same, however, tried different values in all three fields, texthl, linehl and numhl, but to no effect.

**Reply by j0ker70** [reply]
@mfussenegger how to define signs for DapBreakpoint now that sign_define is deprecated?

**Reply by mfussenegger** [reply]
Who said that `sign_define` is deprecated?

**Reply by j0ker70** [reply]
Sorry I was using nvim 0.11. In it, I get a warning that says sign_define in deprecated. It says something like this when I entered `:h deprecated`:

```
• Configuring |diagnostic-signs| using |:sign-define| or |sign_define()|. Use
  the "signs" key of |vim.diagnostic.config()| instead.
```

**Reply by j0ker70** [reply]
You can also check this out:
https://neovim.io/doc/user/deprecated.html

**Reply by mfussenegger** [reply]
That's only specific to `vim.diagnostic.config`. `sign_define` itself is not deprecated

**Reply by j0ker70** [reply]
Can you kindly show an example of how to use `vim.diagnostic.config`'s signs key to define DapBreakpoint?

**Reply by mfussenegger** [reply]
You don't. Sign definitions for nvim-dap are completely unrelated to sign definitions for `vim.diagnostic`.

See the "SIGNS CONFIGURATION" section in `:help dap.txt`

**Comment by josefvn** [comment]
Sweet, managed to get this working, using Lua.

```
vim.highlight.create('DapBreakpoint', { ctermbg=0, guifg='#993939', guibg='#31353f' }, false)
vim.highlight.create('DapLogPoint', { ctermbg=0, guifg='#61afef', guibg='#31353f' }, false)
vim.highlight.create('DapStopped', { ctermbg=0, guifg='#98c379', guibg='#31353f' }, false)

vim.fn.sign_define('DapBreakpoint', { text='', texthl='DapBreakpoint', linehl='DapBreakpoint', numhl='DapBreakpoint' })
vim.fn.sign_define('DapBreakpointCondition', { text='ﳁ', texthl='DapBreakpoint', linehl='DapBreakpoint', numhl='DapBreakpoint' })
vim.fn.sign_define('DapBreakpointRejected', { text='', texthl='DapBreakpoint', linehl='DapBreakpoint', numhl= 'DapBreakpoint' })
vim.fn.sign_define('DapLogPoint', { text='', texthl='DapLogPoint', linehl='DapLogPoint', numhl= 'DapLogPoint' })
vim.fn.sign_define('DapStopped', { text='', texthl='DapStopped', linehl='DapStopped', numhl= 'DapStopped' })
```

Here's what it looks like:

![Screenshot 2022-02-11 at 18 11 45](https://user-images.githubusercontent.com/612695/153637125-d4eba0e4-e650-4eb6-8808-c970618bd42b.png)

Just need to get the virtual text plugin to adopt the same background color.

**Reply by WhiteBlackGoose** [reply]
Nice, thanks!

**Reply by kaykhan** [reply]
> Sweet, managed to get this working, using Lua.
> 
> ```
> vim.highlight.create('DapBreakpoint', { ctermbg=0, guifg='#993939', guibg='#31353f' }, false)
> vim.highlight.create('DapLogPoint', { ctermbg=0, guifg='#61afef', guibg='#31353f' }, false)
> vim.highlight.create('DapStopped', { ctermbg=0, guifg='#98c379', guibg='#31353f' }, false)
> 
> vim.fn.sign_define('DapBreakpoint', { text='', texthl='DapBreakpoint', linehl='DapBreakpoint', numhl='DapBreakpoint' })
> vim.fn.sign_define('DapBreakpointCondition', { text='ﳁ', texthl='DapBreakpoint', linehl='DapBreakpoint', numhl='DapBreakpoint' })
> vim.fn.sign_define('DapBreakpointRejected', { text='', texthl='DapBreakpoint', linehl='DapBreakpoint', numhl= 'DapBreakpoint' })
> vim.fn.sign_define('DapLogPoint', { text='', texthl='DapLogPoint', linehl='DapLogPoint', numhl= 'DapLogPoint' })
> vim.fn.sign_define('DapStopped', { text='', texthl='DapStopped', linehl='DapStopped', numhl= 'DapStopped' })
> ```
> 
> Here's what it looks like:
> 
> ![Screenshot 2022-02-11 at 18 11 45](https://user-images.githubusercontent.com/612695/153637125-d4eba0e4-e650-4eb6-8808-c970618bd42b.png)
> 
> Just need to get the virtual text plugin to adopt the same background color.

When using `vim.highlight.create` i get a error/waring message in neovim 0.8. It is reccomended to use `api.nvim_set_hl` instead although ive not managed to get the port working. Any help is appreciated

```
local set_namespace = vim.api.nvim__set_hl_ns or vim.api.nvim_set_hl_ns
local namespace = vim.api.nvim_create_namespace("dap-hlng")
vim.api.nvim_set_hl(namespace, 'DapBreakpoint', { fg='#eaeaeb', bg='#ffffff' })
vim.api.nvim_set_hl(namespace, 'DapLogPoint', { fg='#eaeaeb', bg='#ffffff' })
vim.api.nvim_set_hl(namespace, 'DapStopped', { fg='#eaeaeb', bg='#ffffff' })

vim.fn.sign_define('DapBreakpoint', { text='', texthl='DapBreakpoint', linehl='DapBreakpoint', numhl='DapBreakpoint' })
vim.fn.sign_define('DapBreakpointCondition', { text='ﳁ', texthl='DapBreakpoint', linehl='DapBreakpoint', numhl='DapBreakpoint' })
vim.fn.sign_define('DapBreakpointRejected', { text='', texthl='DapBreakpoint', linehl='DapBreakpoint', numhl= 'DapBreakpoint' })
vim.fn.sign_define('DapLogPoint', { text='', texthl='DapLogPoint', linehl='DapLogPoint', numhl= 'DapLogPoint' })
vim.fn.sign_define('DapStopped', { text='', texthl='DapStopped', linehl='DapStopped', numhl= 'DapStopped' })
```

**Reply by Strus** [reply]
@kaykhan Following works for me:
```lua
vim.api.nvim_set_hl(0, 'DapBreakpoint', { ctermbg = 0, fg = '#993939', bg = '#31353f' })
vim.api.nvim_set_hl(0, 'DapLogPoint', { ctermbg = 0, fg = '#61afef', bg = '#31353f' })
vim.api.nvim_set_hl(0, 'DapStopped', { ctermbg = 0, fg = '#98c379', bg = '#31353f' })
```

**Comment by lobophf** [comment]
@kaykhan You can try something like this:

```
vim.api.nvim_set_hl(0, "blue",   { fg = "#3d59a1" }) 
vim.api.nvim_set_hl(0, "green",  { fg = "#9ece6a" }) 
vim.api.nvim_set_hl(0, "yellow", { fg = "#FFFF00" }) 
vim.api.nvim_set_hl(0, "orange", { fg = "#f09000" }) 

vim.fn.sign_define('DapBreakpoint',          { text='•', texthl='blue',   linehl='DapBreakpoint', numhl='DapBreakpoint' })
vim.fn.sign_define('DapBreakpointCondition', { text='•', texthl='blue',   linehl='DapBreakpoint', numhl='DapBreakpoint' })
vim.fn.sign_define('DapBreakpointRejected',  { text='•', texthl='orange', linehl='DapBreakpoint', numhl='DapBreakpoint' })
vim.fn.sign_define('DapStopped',             { text='•', texthl='green',  linehl='DapBreakpoint', numhl='DapBreakpoint' })
vim.fn.sign_define('DapLogPoint',            { text='•', texthl='yellow', linehl='DapBreakpoint', numhl='DapBreakpoint' })
```

**Comment by Cruising0904** [comment]
I can't change the color aswell  I just put 🔴 this inside of text it appears with as look as is

**Reply by lobophf** [reply]
@Cruising0904, you can't change the color of this character because it is already red.
However, you can use another one or try the same as I did in [this](https://github.com/mfussenegger/nvim-dap/discussions/355#discussioncomment-4425804) comment:

**Comment by ghost** [comment]
Since some colorschemes might clear the user-defined colors. Here is a improved version of solution.
```
vim.api.nvim_create_autocmd("ColorScheme", {
	pattern = "*",
	group = "UserDefLoadOnce",
	desc = "prevent colorscheme clears self-defined DAP icon colors.",
	callback = function()
		vim.api.nvim_set_hl(0, 'DapBreakpoint', { ctermbg = 0, fg = '#993939' })
		vim.api.nvim_set_hl(0, 'DapLogPoint', { ctermbg = 0, fg = '#61afef' })
		vim.api.nvim_set_hl(0, 'DapStopped', { ctermbg = 0, fg = '#98c379' })
	end
})

vim.fn.sign_define('DapBreakpoint', { text=' ', texthl='DapBreakpoint' })
vim.fn.sign_define('DapBreakpointCondition', { text=' ﳁ', texthl='DapBreakpoint' })
vim.fn.sign_define('DapBreakpointRejected', { text=' ', texthl='DapBreakpoint' })
vim.fn.sign_define('DapLogPoint', { text=' ', texthl='DapLogPoint' })
vim.fn.sign_define('DapStopped', { text=' ', texthl='DapStopped' }
```

**Reply by shmerl** [reply]
I'm trying this method (I think in my case the theme also prevents that as above), but I get an error:

```
E5108: Error executing lua [string ":lua"]:5: Invalid 'group': 'UserDefLoadOnce'
stack traceback:
        [C]: in function 'nvim_create_autocmd'
        [string ":lua"]:5: in main chunk
```

**Reply by shmerl** [reply]
Hm, removing `group = "UserDefLoadOnce",` seems to work.

**Reply by shmerl** [reply]
I found a working method that honors current color scheme background color for SignColumn without overwriting it:

```lua
vim.api.nvim_create_autocmd("ColorScheme", {
  pattern = "*",
  desc = "Prevent colorscheme clearing self-defined DAP marker colors",
  callback = function()
      -- Reuse current SignColumn background (except for DapStoppedLine)
      local sign_column_hl = vim.api.nvim_get_hl(0, { name = 'SignColumn' })
      -- if bg or ctermbg aren't found, use bg = 'bg' (which means current Normal) and ctermbg = 'Black'
      -- convert to 6 digit hex value starting with #
      local sign_column_bg = (sign_column_hl.bg ~= nil) and ('#%06x'):format(sign_column_hl.bg) or 'bg'
      local sign_column_ctermbg = (sign_column_hl.ctermbg ~= nil) and sign_column_hl.ctermbg or 'Black'

      vim.api.nvim_set_hl(0, 'DapStopped', { fg = '#00ff00', bg = sign_column_bg, ctermbg = sign_column_ctermbg })
      vim.api.nvim_set_hl(0, 'DapStoppedLine', { bg = '#2e4d3d', ctermbg = 'Green' })
      vim.api.nvim_set_hl(0, 'DapBreakpoint', { fg = '#c23127', bg = sign_column_bg, ctermbg = sign_column_ctermbg })
      vim.api.nvim_set_hl(0, 'DapBreakpointRejected', { fg = '#888ca6', bg = sign_column_bg, ctermbg = sign_column_ctermbg })
      vim.api.nvim_set_hl(0, 'DapLogPoint', { fg = '#61afef', bg = sign_column_bg, ctermbg = sign_column_ctermbg })
  end
})

-- reload current color scheme to pick up colors override if it was set up in a lazy plugin definition fashion
vim.cmd.colorscheme(vim.g.colors_name)
```

**Reply by shmerl** [reply]
See also an improved version of this below, without calling `vim.cmd.colorscheme`.

**Comment by shmerl** [comment]
After switching to using plugin management with lazy.nvim, no method of setting colors works now with any color scheme.

Before I was using that `vim.api.nvim_create_autocmd` referenced above (when trying it with plug.vim) and it worked. Did anyone try that with lazy.nvim?

UPDATE:

Actually, since update to neovim v0.9.5 it's not working at all even with plug.vim. I'll try older nvim.

UPDATE 2:

OK, my bad, something else is wrong with my set up. Even nvim  v0.9.4 isn't working. I'll try to narrow it down.

UPDATE 3:

OK, I see what's going on. Setting of the colorscheme should happen **after** that autocmd is defined, otherwise it won't work.

Though, lazy.nvim recommends setting of the colorscheme to have highest priority. I suppose that's not really suitable for this case. I just split that autocmd into a separate module from the plugin configuration to make sure it works and load it before lazy.nvim itself.

**Comment by shmerl** [comment]
Another small update. I managed to enforce the settings with autocmd without explicitly setting colorscheme. The trick was to use augroup and `nvim_exec_autocmds`. My current example:

```lua
-- augroup to be able to trigger the autocommand explicitly for the first time
vim.api.nvim_create_augroup("dap_colors", {})

-- autocmd to enforce colors settings on any color scheme change
vim.api.nvim_create_autocmd("ColorScheme", {
   group = "dap_colors",
   pattern = "*",
   desc = "Set DAP marker colors and prevent color theme from resetting them",
   callback = function()
      -- Reuse current SignColumn background (except for DapStoppedLine)
      local sign_column_hl = vim.api.nvim_get_hl(0, { name = 'SignColumn' })
      -- if bg or ctermbg aren't found, use bg = 'bg' (which means current Normal) and ctermbg = 'Black'
      -- convert to 6 digit hex value starting with #
      local sign_column_bg = (sign_column_hl.bg ~= nil) and ('#%06x'):format(sign_column_hl.bg) or 'bg'
      local sign_column_ctermbg = (sign_column_hl.ctermbg ~= nil) and sign_column_hl.ctermbg or 'Black'

      vim.api.nvim_set_hl(0, 'DapStopped', { fg = '#00ff00', bg = sign_column_bg, ctermbg = sign_column_ctermbg })
      vim.api.nvim_set_hl(0, 'DapStoppedLine', { bg = '#2e4d3d', ctermbg = 'Green' })
      vim.api.nvim_set_hl(0, 'DapBreakpoint', { fg = '#c23127', bg = sign_column_bg, ctermbg = sign_column_ctermbg })
      vim.api.nvim_set_hl(0, 'DapBreakpointRejected', { fg = '#888ca6', bg = sign_column_bg, ctermbg = sign_column_ctermbg })
      vim.api.nvim_set_hl(0, 'DapLogPoint', { fg = '#61afef', bg = sign_column_bg, ctermbg = sign_column_ctermbg })
   end
})

-- executing the settings explicitly for the first time
vim.api.nvim_exec_autocmds("ColorScheme", { group = "dap_colors" })
```
